### PR TITLE
ddl: add recover for run ddl job

### PR DESCRIPTION
**Pull request by crazycs520**
### What problem does this PR solve? 
Fix ddl worker run ddl job panic.

```sql
[2019/06/28 15:21:54.379 +08:00] [INFO] [ddl_worker.go:474] ["[ddl] run DDL job"] [worker="worker 1, tp general"] [job="ID:48, Type:create table, State:none, SchemaState:none, SchemaID:1, TableID:47, RowCount:0, ArgLen:0, start time: 2019-06-28 15:19:12.653 +0800 CST, Err:<nil>, ErrCount:0, SnapshotVersion:0"]
[2019/06/28 15:21:54.379 +08:00] [ERROR] [ddl.go:441] ["[ddl] DDL worker meet panic"] [worker="worker 1, tp general"] [ID=8c8b0d58-4238-4f21-bd15-6915af71a983] [stack="github.com/pingcap/tidb/ddl.(*ddl).start.func2
	/Users/cs/code/goread/src/github.com/pingcap/tidb/ddl/ddl.go:441
github.com/pingcap/tidb/util.WithRecovery.func1
	/Users/cs/code/goread/src/github.com/pingcap/tidb/util/misc.go:72
runtime.gopanic
	/usr/local/go1.12/src/runtime/panic.go:522
runtime.panicmem
	/usr/local/go1.12/src/runtime/panic.go:82
runtime.sigpanic
	/usr/local/go1.12/src/runtime/signal_unix.go:390
github.com/pingcap/tidb/infoschema.(*Handle).Get
	/Users/cs/code/goread/src/github.com/pingcap/tidb/infoschema/infoschema.go:302
github.com/pingcap/tidb/ddl.onCreateTable
	/Users/cs/code/goread/src/github.com/pingcap/tidb/ddl/table.go:56
github.com/pingcap/tidb/ddl.(*worker).runDDLJob
	/Users/cs/code/goread/src/github.com/pingcap/tidb/ddl/ddl_worker.go:499
github.com/pingcap/tidb/ddl.(*worker).handleDDLJobQueue.func1
	/Users/cs/code/goread/src/github.com/pingcap/tidb/ddl/ddl_worker.go:402
github.com/pingcap/tidb/kv.RunInNewTxn
	/Users/cs/code/goread/src/github.com/pingcap/tidb/kv/txn.go:50
github.com/pingcap/tidb/ddl.(*worker).handleDDLJobQueue
	/Users/cs/code/goread/src/github.com/pingcap/tidb/ddl/ddl_worker.go:365
github.com/pingcap/tidb/ddl.(*worker).start
	/Users/cs/code/goread/src/github.com/pingcap/tidb/ddl/ddl_worker.go:137
github.com/pingcap/tidb/ddl.(*ddl).start.func1
	/Users/cs/code/goread/src/github.com/pingcap/tidb/ddl/ddl.go:438
github.com/pingcap/tidb/util.WithRecovery
	/Users/cs/code/goread/src/github.com/pingcap/tidb/util/misc.go:80"]
[2019/06/28 15:21:54.380 +08:00] [ERROR] [misc.go:75] ["panic in the recoverable goroutine"] [r="\"invalid memory address or nil pointer dereference\""] ["stack trace"="github.com/pingcap/tidb/util.WithRecovery.func1
	/Users/cs/code/goread/src/github.com/pingcap/tidb/util/misc.go:77
runtime.gopanic
	/usr/local/go1.12/src/runtime/panic.go:522
runtime.panicmem
	/usr/local/go1.12/src/runtime/panic.go:82
runtime.sigpanic
	/usr/local/go1.12/src/runtime/signal_unix.go:390
github.com/pingcap/tidb/infoschema.(*Handle).Get
	/Users/cs/code/goread/src/github.com/pingcap/tidb/infoschema/infoschema.go:302
github.com/pingcap/tidb/ddl.onCreateTable
	/Users/cs/code/goread/src/github.com/pingcap/tidb/ddl/table.go:56
github.com/pingcap/tidb/ddl.(*worker).runDDLJob
	/Users/cs/code/goread/src/github.com/pingcap/tidb/ddl/ddl_worker.go:499
github.com/pingcap/tidb/ddl.(*worker).handleDDLJobQueue.func1
	/Users/cs/code/goread/src/github.com/pingcap/tidb/ddl/ddl_worker.go:402
github.com/pingcap/tidb/kv.RunInNewTxn
	/Users/cs/code/goread/src/github.com/pingcap/tidb/kv/txn.go:50
github.com/pingcap/tidb/ddl.(*worker).handleDDLJobQueue
	/Users/cs/code/goread/src/github.com/pingcap/tidb/ddl/ddl_worker.go:365
github.com/pingcap/tidb/ddl.(*worker).start
	/Users/cs/code/goread/src/github.com/pingcap/tidb/ddl/ddl_worker.go:137
github.com/pingcap/tidb/ddl.(*ddl).start.func1
	/Users/cs/code/goread/src/github.com/pingcap/tidb/ddl/ddl.go:438
github.com/pingcap/tidb/util.WithRecovery
	/Users/cs/code/goread/src/github.com/pingcap/tidb/util/misc.go:80"] [stack="github.com/pingcap/tidb/util.WithRecovery.func1
	/Users/cs/code/goread/src/github.com/pingcap/tidb/util/misc.go:75
runtime.gopanic
	/usr/local/go1.12/src/runtime/panic.go:522
runtime.panicmem
	/usr/local/go1.12/src/runtime/panic.go:82
runtime.sigpanic
	/usr/local/go1.12/src/runtime/signal_unix.go:390
github.com/pingcap/tidb/infoschema.(*Handle).Get
	/Users/cs/code/goread/src/github.com/pingcap/tidb/infoschema/infoschema.go:302
github.com/pingcap/tidb/ddl.onCreateTable
	/Users/cs/code/goread/src/github.com/pingcap/tidb/ddl/table.go:56
github.com/pingcap/tidb/ddl.(*worker).runDDLJob
	/Users/cs/code/goread/src/github.com/pingcap/tidb/ddl/ddl_worker.go:499
github.com/pingcap/tidb/ddl.(*worker).handleDDLJobQueue.func1
	/Users/cs/code/goread/src/github.com/pingcap/tidb/ddl/ddl_worker.go:402
github.com/pingcap/tidb/kv.RunInNewTxn
	/Users/cs/code/goread/src/github.com/pingcap/tidb/kv/txn.go:50
github.com/pingcap/tidb/ddl.(*worker).handleDDLJobQueue
	/Users/cs/code/goread/src/github.com/pingcap/tidb/ddl/ddl_worker.go:365
github.com/pingcap/tidb/ddl.(*worker).start
	/Users/cs/code/goread/src/github.com/pingcap/tidb/ddl/ddl_worker.go:137
github.com/pingcap/tidb/ddl.(*ddl).start.func1
	/Users/cs/code/goread/src/github.com/pingcap/tidb/ddl/ddl.go:438
github.com/pingcap/tidb/util.WithRecovery
	/Users/cs/code/goread/src/github.com/pingcap/tidb/util/misc.go:80"]
```

### What is changed and how it works?
* Add a recover in `runDDLJob` function.
* check nil point.

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change

Side effects

Related changes

 - Need to cherry-pick to the release branch

### CODE REVIEW

File: ddl/table.go [hunk 1]
```diff
@@ -773,7 +773,7 @@ func onModifyTableCharsetAndCollate(t *meta.Meta, job *model.Job) (ver int64, _
 
 func checkTableNotExists(d *ddlCtx, t *meta.Meta, schemaID int64, tableName string) error {
 	// d.infoHandle maybe nil in some test.
-	if d.infoHandle == nil {
+	if d.infoHandle == nil || d.infoHandle.IsNil() {
```
Comment (zimulala) [inline ddl/table.go]: Could we rename this function name? `IsNil` made me think that it is used to judge whether the value of `infoHandle` is empty.

Comment (crazycs520) [reply ddl/table.go]: How about `Isvalid`

File: ddl/serial_test.go [hunk 2]
```diff
@@ -429,3 +429,17 @@ func (s *testSerialSuite) TestTableLocksEnable(c *C) {
 	tk.MustExec("lock tables t1 write")
 	checkTableLock(c, tk.Se, "test", "t1", model.TableLockNone)
 }
+
+// TestRunDDLJobPanic tests recover panic when run ddl job panic.
+func (s *testSerialSuite) TestRunDDLJobPanic(c *C) {
```
Comment (zimulala) [inline ddl/serial_test.go]: Could we put it to `ddl/failtest`?

File: ddl/ddl_worker.go [hunk 3]
```diff
@@ -471,6 +472,22 @@ func chooseLeaseTime(t, max time.Duration) time.Duration {
 
 // runDDLJob runs a DDL job. It returns the current schema version in this transaction and the error.
 func (w *worker) runDDLJob(d *ddlCtx, t *meta.Meta, job *model.Job) (ver int64, err error) {
+	defer func() {
+		r := recover()
```
Comment (zimulala) [inline ddl/ddl_worker.go]: We have added `recover` to the worker's `start`. Is it repeated？

File: ddl/serial_test.go [hunk 4]
```diff
@@ -429,3 +429,17 @@ func (s *testSerialSuite) TestTableLocksEnable(c *C) {
...
+	defer func() {
+		c.Assert(failpoint.Disable("github.com/pingcap/tidb/ddl/mockPanicInRunDDLJob"), IsNil)
+	}()
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec("use test")
+	tk.MustExec("drop table if exists t")
+	c.Assert(failpoint.Enable("github.com/pingcap/tidb/ddl/mockPanicInRunDDLJob", `1*return(true)`), IsNil)
```
Comment (amyangfei) [inline ddl/serial_test.go]:
```suggestion
	c.Assert(failpoint.Enable("github.com/pingcap/tidb/ddl/mockPanicInRunDDLJob", `1*panic("panic test")`), IsNil)
```
and inject code with
```
failpoint.Inject("mockPanicInRunDDLJob", nil)
```
is simpler

File: ddl/serial_test.go (hunk 2 above)
Comment (crazycs520) [reply ddl/serial_test.go]: done.

File: ddl/table.go (hunk 1 above)
Comment (zimulala) [reply ddl/table.go]: I think `IsValid` is OK. But I see other one using `Valid` in our TiDB.

File: ddl/ddl_worker.go (hunk 3 above)
Comment (winkyao) [reply ddl/ddl_worker.go]: not enough, what if all worker panic?

File: ddl/serial_test.go (hunk 4 above)
Comment (crazycs520) [reply ddl/serial_test.go]: Great, thanks.

File: ddl/ddl_worker.go (hunk 3 above)
Comment (crazycs520) [reply ddl/ddl_worker.go]: No reapeat，`recover` in `start` will cause ddl worker already  exit, and then there is no ddl worker in this tidb server.

File: ddl/ddl_worker.go [hunk 5]
```diff
@@ -471,6 +472,18 @@ func chooseLeaseTime(t, max time.Duration) time.Duration {
...
```
Comment (zimulala) [inline ddl/ddl_worker.go]: Could we use `WithRecovery`？

File: ddl/ddl_worker.go [hunk 6]
```diff
@@ -471,6 +472,18 @@ func chooseLeaseTime(t, max time.Duration) time.Duration {
...
+		if r != nil {
+			logutil.BgLogger().Error("run ddl job panic",
+				zap.Reflect("r", r),
+				zap.Stack("stack trace"))
+			job.State = model.JobStateCancelling
```
Comment (winkyao) [inline ddl/ddl_worker.go]: add some comments

File: infoschema/infoschema.go [hunk 7]
```diff
@@ -303,6 +303,11 @@ func (h *Handle) Get() InfoSchema {
 	return schema
 }
 
+// IsValid uses to check whether handle value is nil.
+func (h *Handle) IsValid() bool {
```
Comment (winkyao) [inline infoschema/infoschema.go]: IsValid? I think it means IsNotValid.

Comment (bb7133) [reply infoschema/infoschema.go]: Please also update the comment

File: ddl/ddl_worker.go (hunk 5 above)
Comment (crazycs520) [reply ddl/ddl_worker.go]: done.

File: ddl/ddl_worker.go (hunk 6 above)
Comment (crazycs520) [reply ddl/ddl_worker.go]: done.

File: infoschema/infoschema.go (hunk 7 above)
Comment (crazycs520) [reply infoschema/infoschema.go]: done.

**Comment by bb7133** [comment]
LGTM

**Pull request merged**